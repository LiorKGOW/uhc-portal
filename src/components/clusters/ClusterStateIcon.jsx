import React from 'react';
import PropTypes from 'prop-types';
import {
  Icon,
} from 'patternfly-react';

function ClusterStateIcon(props) {
  const {
    clusterState
  } = props;

  let icon;
  switch (clusterState) {
    case 'Installing':
      icon = 'maintenance';
      break;
    case 'Error':
      icon = 'error-circle-o';
      break;
    case 'OK':
      icon = 'ok';
      break;
    default:
      icon = 'unknown';
  }
  return (
    <Icon name={icon} type="pf" />);
}

ClusterStateIcon.propTypes = {
  clusterState: PropTypes.string.isRequired,
};

export default ClusterStateIcon;