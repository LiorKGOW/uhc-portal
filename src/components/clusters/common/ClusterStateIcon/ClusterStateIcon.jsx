// ClusterStateIcon matches a cluster state from the API to the matching icon

import React from 'react';
import PropTypes from 'prop-types';
import {
  CheckCircleIcon,
  DisconnectedIcon,
  UnknownIcon,
  ExclamationCircleIcon,
  InProgressIcon,
  BanIcon,
  FolderOpenIcon,
} from '@patternfly/react-icons';
// need to disable eslint for the react tokens because it's silly - it warns about these names
// eslint-disable-next-line camelcase
import { global_danger_color_100, global_success_color_100 } from '@patternfly/react-tokens';
import { Spinner } from '@patternfly/react-core';
import clusterStates from '../clusterStates';

function ClusterStateIcon(props) {
  const { clusterState, animated } = props;

  const iconProps = {
    className: 'clusterstate',
    size: 'sm',
  };

  // Icons from http://openshift.github.io/openshift-origin-design/web-console/4.0-designs/status/status
  switch (clusterState) {
    case clusterStates.PENDING:
    case clusterStates.INSTALLING:
    case clusterStates.UPDATING:
      if (animated) {
        return <Spinner {...iconProps} />;
      }
      return <InProgressIcon {...iconProps} />;
    case clusterStates.DISCONNECTED:
      return <DisconnectedIcon {...iconProps} />;
    case clusterStates.READY:
      return <CheckCircleIcon color={global_success_color_100.value} {...iconProps} />;
    case clusterStates.UNINSTALLING:
      if (animated) {
        return <Spinner {...iconProps} />;
      }
      return <InProgressIcon {...iconProps} />;
    case clusterStates.ERROR:
      return <ExclamationCircleIcon color={global_danger_color_100.value} {...iconProps} />;
    case clusterStates.DEPROVISIONED:
      return <BanIcon {...iconProps} />;
    case clusterStates.ARCHIVED:
      return <FolderOpenIcon {...iconProps} />;
    default:
      return <UnknownIcon {...iconProps} />;
  }
}

ClusterStateIcon.propTypes = {
  clusterState: PropTypes.string.isRequired,
  animated: PropTypes.bool,
};
ClusterStateIcon.defaultProps = {
  animated: false,
};
export default ClusterStateIcon;
