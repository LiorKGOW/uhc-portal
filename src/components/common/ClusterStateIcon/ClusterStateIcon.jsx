// ClusterStateIcon matches a cluster state from the API to the matching icon

import React from 'react';
import PropTypes from 'prop-types';
import { Icon } from 'patternfly-react';
import clusterStates from '../../../common/clusterStates';

function ClusterStateIcon(props) {
  const { clusterState } = props;

  let icon = { type: 'pf' };
  // Icons from http://openshift.github.io/openshift-origin-design/web-console/4.0-designs/status/status
  switch (clusterState) {
    case clusterStates.PENDING:
      icon = { type: 'fa', name: 'hourglass-half' };
      break;
    case clusterStates.INSTALLING:
      icon.name = 'in-progress';
      break;
    case clusterStates.ERROR:
      icon.name = 'error-circle-o';
      break;
    case clusterStates.READY:
      icon.name = 'ok';
      break;
    case clusterStates.UNINSTALLING:
      icon = { type: 'fa', name: 'ban' };
      break;
    case clusterStates.PATCHING:
      icon = { type: 'fa', name: 'hourglass-half' };
      break;
    default:
      icon.name = 'unknown';
  }
  // patternfly bug workaround: pf icons ignore the `size` prop.
  // Specifying className='fa-lg' makes them larger too.
  return (
    <Icon className="fa-lg clusterstate" {...icon} />);
}

ClusterStateIcon.propTypes = {
  clusterState: PropTypes.string.isRequired,
};

export default ClusterStateIcon;
