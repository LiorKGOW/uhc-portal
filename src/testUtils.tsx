import React from 'react';
import { Provider } from 'react-redux';
import { createStore, Store } from 'redux';
import { render, RenderOptions } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import '@testing-library/jest-dom';

import { toHaveNoViolations, axe } from 'jest-axe';

import { createBrowserHistory } from 'history';
import { store } from './redux/store';
import { reduxReducers } from './redux/reducers';
import * as restrictedEnv from './restrictedEnv';

const history = createBrowserHistory();

export const TestWrapper = ({
  children,
  initialStore,
}: {
  children: React.ReactNode;
  initialStore?: Store;
}) => <Provider store={initialStore || store}>{children}</Provider>;

const renderWithState = (
  ui: React.ReactElement,
  options?: Omit<RenderOptions, 'queries'>,
  initialState?: any, // setting to any because only a partial state structure may be sent
) => {
  const newStore = initialState ? createStore(reduxReducers(history), initialState) : store;

  const Wrapper = ({ children }: { children: React.ReactNode }) => (
    <TestWrapper initialStore={newStore}>{children}</TestWrapper>
  );
  return { ...render(ui, { wrapper: Wrapper, ...options }), user: userEvent.setup() };
};

export * from '@testing-library/react';
export { default as userEvent } from '@testing-library/user-event';

// Type not exported in the library
export type UserEventType = {
  type: (
    field: HTMLElement,
    typeValue: string,
    { initialSelectionStart }: { initialSelectionStart: number },
  ) => void;
  clear: (field: HTMLElement) => void;
};

export { renderWithState as render };

/* ***** Items outside of React Test Library ************ */
expect.extend(toHaveNoViolations);

export const checkAccessibility = async (container: HTMLElement | string) => {
  const results = await axe(container);
  expect(results).toHaveNoViolations();
};

export const insightsMock = () => {
  global.insights = {
    chrome: {
      ...global.insights.chrome,
      on: () => () => {},
      appNavClick: () => {},
      auth: {
        getUser: () => Promise.resolve({ data: {} }),
        getToken: () => Promise.resolve(),
        getOfflineToken: () => Promise.resolve({ data: { refresh_token: 'hello' } }),
      },
    },
  };
};

export const mockRestrictedEnv = () => {
  const mock = jest.spyOn(restrictedEnv, 'isRestrictedEnv');
  mock.mockReturnValue(false);
  return mock;
};
