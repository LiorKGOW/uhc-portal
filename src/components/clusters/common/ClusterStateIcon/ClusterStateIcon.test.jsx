import React from 'react';
import { render } from '~/testUtils';

import ClusterStateIcon from './ClusterStateIcon';

// NOTE there is no way to really know what icon is displayed, so we are depending on a data attribute

describe('<ClusterStateIcon />', () => {
  it.each([
    ['unknown', 'whatever'],
    ['unknown', 'warning'],
    ['inprogress', 'pending'],
    ['inprogress', 'installing'],
    ['inprogress', 'uninstalling'],
    ['exclamation', 'error'],
    ['check', 'ready'],
    [undefined, 'unknown'],
  ])('renders %p icon for %p state', (icon, state) => {
    const { container } = render(<ClusterStateIcon clusterState={state} />);
    expect(container.querySelector('svg')).toHaveAttribute('data-icon-type', icon);
  });
});
