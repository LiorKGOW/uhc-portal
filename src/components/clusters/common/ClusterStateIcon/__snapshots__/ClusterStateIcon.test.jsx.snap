// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ClusterStateIcon /> renders correctly for error status 1`] = `
<Icon
  className="fa-lg clusterstate "
  name="error-circle-o"
  type="pf"
/>
`;

exports[`<ClusterStateIcon /> renders correctly for installing status 1`] = `
<Icon
  className="fa-lg clusterstate "
  name="in-progress"
  type="pf"
/>
`;

exports[`<ClusterStateIcon /> renders correctly for patching status 1`] = `
<Icon
  className="fa-lg clusterstate "
  name="hourglass-half"
  type="fa"
/>
`;

exports[`<ClusterStateIcon /> renders correctly for pending status 1`] = `
<Icon
  className="fa-lg clusterstate "
  name="hourglass-half"
  type="fa"
/>
`;

exports[`<ClusterStateIcon /> renders correctly for ready status 1`] = `
<Icon
  className="fa-lg clusterstate "
  name="ok"
  type="pf"
/>
`;

exports[`<ClusterStateIcon /> renders correctly for uninstalling status 1`] = `
<Icon
  className="fa-lg clusterstate "
  name="ban"
  type="fa"
/>
`;

exports[`<ClusterStateIcon /> renders correctly for unknown status 1`] = `
<Icon
  className="fa-lg clusterstate "
  name="unknown"
  type="pf"
/>
`;

exports[`<ClusterStateIcon /> renders correctly for warning status 1`] = `
<Icon
  className="fa-lg clusterstate "
  name="warning-triangle-o"
  type="pf"
/>
`;
